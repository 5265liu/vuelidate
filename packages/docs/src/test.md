# Live test components

## As Object
<as-object></as-object>


## As Composition
