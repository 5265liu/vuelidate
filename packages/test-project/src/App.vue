<template>
  <div id="app">
    <nested-validations />
  </div>
</template>

<script>
import NestedValidations from '@/components/NestedValidations'

export default {
  name: 'App',
  components: {
    NestedValidations
  }
}
</script>

<style>
#app {
  font-family: Avenir, Helvetica, Arial, sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  color: #2c3e50;
  margin-top: 60px;
}
</style>
