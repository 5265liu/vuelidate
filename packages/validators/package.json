{
  "name": "@vuelidate/validators",
  "version": "0.0.1",
  "description": "Validators for Vuelidate",
  "main": "dist/index.js",
  "module": "dist/index.esm.js",
  "repository": "https://github.com/vuelidate/vuelidate",
  "license": "MIT",
  "scripts": {
    "build": "bili src src/raw --format esm --format cjs",
    "lint": "eslint src"
  },
  "devDependencies": {
    "bili": "^4.8.1",
    "eslint-plugin-jest": "^23.0.2",
    "jest": "^24.9.0"
  },
  "sideEffects": [
    "./src/utils/common.js"
  ]
}
