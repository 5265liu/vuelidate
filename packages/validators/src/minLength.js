import minLength from './raw/minLength'

export default length => ({
  $validator: minLength,
  $message: ({ $params }) => `This field should be at least ${$params.length} long.`,
  $params: { length }
})
