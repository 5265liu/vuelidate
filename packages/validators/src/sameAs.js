import sameAs from './raw/sameAs'

export default equalTo => ({
  $validator: sameAs,
  $message: 'This field should be equal as the other field.',
  $params: { equalTo }
})
