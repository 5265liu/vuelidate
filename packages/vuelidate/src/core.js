import { isFunction, isPromise, unwrap, unwrapObj } from './utils'
import { computed, reactive, ref, watch } from '@vue/composition-api'

function sortValidations (validations) {
  const validationKeys = Object.keys(validations)

  let rules = {}
  let nested = {}
  let config = {}

  validationKeys.forEach(key => {
    const v = validations[key]

    switch (true) {
      // If it is already normalized, use it
      case isFunction(v.$validator):
        rules[key] = v
        break
      // If it is just a function, normalize it first
      // into { $validator: <Fun> }
      case isFunction(v):
        rules[key] = { $validator: v }
        break
      // Catch $-prefixed properties as config
      case key.startsWith('$'):
        config[key] = v
        break
      // If it doesn’t match any of the above,
      // treat as nested state property
      default:
        nested[key] = v
    }
  })

  return { rules, nested, config }
}

function callRule (rule, value) {
  const v = unwrap(value)
  return rule(v)
}

function createComputedResult (rule, model) {
  return computed(() => {
    const result = callRule(rule, model)
    return result.$invalid !== undefined
      ? result.$invalid
      : !result
  })
}

function createAsyncResult (rule, model, initResult, $pending) {
  const $invalid = ref(true)

  $pending.value = true

  initResult.then(data => {
    $pending.value = false
    $invalid.value = !data
  })

  watch(
    model,
    modelValue => {
      const ruleResult = callRule(rule, modelValue)

      $pending.value = true
      $invalid.value = true

      ruleResult
        .then(data => {
          $pending.value = false
          $invalid.value = !data
        })
        .catch(() => {
          $pending.value = false
          $invalid.value = true
        })
    },
    { lazy: true }
  )

  return $invalid
}

function createValidatorResult (rule, model) {
  const ruleResult = callRule(rule.$validator, model)

  const $pending = ref(false)
  const $params = rule.$params
  const $invalid = isPromise(ruleResult)
    ? createAsyncResult(
      rule.$validator,
      model,
      ruleResult,
      $pending
    )
    : createComputedResult(rule.$validator, model)

  const message = rule.$message
  const $message = isFunction(message)
    ? computed(() => message(
      unwrapObj({
        $pending,
        $invalid,
        $params,
        $model: model
      })
    ))
    : message

  return {
    $message,
    $params,
    $pending,
    $invalid
  }
}

function createValidationResults (rules, state, key, parentKey) {
  const ruleKeys = Object.keys(rules)
  if (!ruleKeys.length) return {}

  const $dirty = ref(false)

  let result = {
    $dirty,
    $touch: () => { $dirty.value = true },
    $reset: () => { $dirty.value = false },
    $pending: ref(false)
  }

  ruleKeys.forEach(ruleKey => {
    result[ruleKey] = createValidatorResult(
      rules[ruleKey],
      state[key]
    )
  })

  result.$invalid = computed(() =>
    ruleKeys.some(ruleKey => result[ruleKey].$invalid)
  )

  result.$error = computed(() =>
    result.$invalid.value && $dirty.value
  )

  result.$errors = computed(() => ruleKeys
    .filter(ruleKey => unwrap(result[ruleKey]).$invalid)
    .map(ruleKey => {
      const res = result[ruleKey]
      return {
        $property: parentKey ? `${parentKey}.${key}` : key,
        $validator: ruleKey,
        $message: res.$message,
        $params: res.$params,
        $pending: res.$pending
      }
    })
  )

  return result
}

function collectNestedValidationResults (validations, state, key) {
  const nestedValidationKeys = Object.keys(validations)

  if (!nestedValidationKeys.length) return {}

  return nestedValidationKeys.reduce((results, nestedKey) => {
    const nestedState = key ? state[key] : state

    results[nestedKey] = setValidations({
      validations: validations[nestedKey],
      state: nestedState,
      key: nestedKey,
      parentKey: key
    })
    return results
  }, {})
}

function createMetaFields (results, nestedResults) {
  const $dirty = ref(false)

  const $errors = computed(() => {
    const modelErrors = unwrap(results.$errors) || []

    const nestedErrors = Object.values(nestedResults)
      .filter(result => result.$errors.length)
      .reduce((errors, result) => {
        return errors.concat(...result.$errors)
      }, [])

    return modelErrors.concat(nestedErrors)
  })

  const $invalid = computed(() =>
    Object.values(nestedResults).some(r => r.$invalid) ||
    unwrap(results.$invalid) ||
    false
  )

  const $anyDirty = computed(() =>
    Object.values(nestedResults).some(r => r.$dirty)
  )

  const $error = computed(() => ($invalid.value && $dirty.value) || false)

  return {
    $dirty,
    $errors,
    $invalid,
    $anyDirty,
    $error
  }
}

/**
 * Main Vuelidate bootstrap function.
 * Used both for Composition API in `setup` and for Global App usage
 * @param {Object} validations
 * @param {Object} state
 * @param {String} key - Current state key. Used when being called on nested items
 * @param parentKey
 * @param childResults
 * @return {UnwrapRef<{$anyDirty: *, $error: *, $pending?: (), $invalid: *, $errors: *, $touch?: (), $dirty: *, $reset?: ()}>}
 */
export function setValidations ({ validations, state, key, parentKey, childResults }) {
  // Sort out the validation object into:
  // – rules = validators for current state tree fragment
  // — nested = nested state fragments keys that might contain more validators
  // – config = configuration properties that affect this state fragment
  const { rules, nested, config } = sortValidations(validations)

  // Use rules for the current state fragment and validate it
  const results = createValidationResults(rules, state, key, parentKey)

  // Use nested keys to repeat the process
  // *WARN*: This is recursive
  const nestedResults = collectNestedValidationResults(nested, state, key)

  // Collect and merge this level validation results
  // with all nested validation results
  const {
    $dirty,
    $errors,
    $invalid,
    $anyDirty,
    $error
  } = createMetaFields(results, nestedResults)

  const $model = computed({
    get: () => unwrap(state[key]),
    set: val => {
      $dirty.value = true
      state[key].value = val
    }
  })

  if (config.$autoDirty) {
    watch(
      state[key],
      () => { $dirty.value = true },
      { lazy: true }
    )
  }

  return reactive({
    ...results,
    // NOTE: The order here is very important, since we want to override
    // some of the *results* meta fields with the collective version of it
    // that includes the results of nested state validation results
    ...(key ? { $model } : {}),
    $dirty,
    $error,
    $errors,
    $invalid,
    $anyDirty,
    ...nestedResults
  })
}
