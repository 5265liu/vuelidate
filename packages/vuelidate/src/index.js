import { provide, inject, ref, computed, reactive, watch } from 'vue'
import { isFunction } from './utils'
import { setValidations } from './core'

const VuelidateSymbol = Symbol('vuelidate')

/**
 * Composition API compatible Vuelidate
 * Use inside the `setup` lifecycle hook
 * @param {Object} validations - Validations Object
 * @param {Object} state - State object
 * @param {String} registerAs
 * @return {UnwrapRef<*>}
 */
export default function useVuelidate (validations, state, registerAs) {
  const childResultsRaw = {}
  const childResultsKeys = ref([])
  const childResults = computed(() => childResultsKeys.value.reduce((results, key) => {
    results[key] = childResultsRaw[key]
    return results
  }, {}))
  const injectToParent = inject(VuelidateSymbol, () => {})
  provide(VuelidateSymbol, injectChildResults)

  function injectChildResults (results, key) {
    childResultsRaw[key] = results
    childResultsKeys.value.push(key)
  }

  const validationResults = setValidations({
    validations,
    state,
    childResults
  })

  if (registerAs) {
    injectToParent(validationResults, registerAs)
  }

  if (registerAs && childResultsKeys.value.length) {
    return reactive({
      ...validationResults,
      ...childResults
    })
  } else {
    return reactive(validationResults)
  }
}

/**
 * Vuelidate mixin, used to attach Vuelidate only to specified components
 * Relies on `validations` option to be defined on component instance
 * @type {ComponentOptions}
 */
export const VuelidateMixin = {
  beforeCreate () {
    const options = this.$options
    if (!options.validations) return

    const validations = isFunction(options.validations)
      ? options.validations.call(this)
      : options.validations

    if (!options.computed) options.computed = {}
    if (options.computed.$v) return

    options.computed.$v = () => setValidations({
      validations,
      state: this
    })
  }
}

/**
 * Default way to install Vuelidate globally for entire app.
 * @param {Vue} app
 */
export function VuelidatePlugin (app) {
  app.mixin(VuelidateMixin)
}
