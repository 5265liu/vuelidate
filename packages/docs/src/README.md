# Getting started


Vuelidate 2 is a simple, but powerful, lightweight model-based validation for Vue.js.

It supports Vue 3

## Installation

Package is installable via npm

```bash
npm install vuelidate --save
```

## Basic usage

You can import the library and `use` as a Vue plugin to enable the functionality globally on all components containing validation configuration.

```js
import Vue from 'vue'
import Vuelidate from 'vuelidate'
Vue.use(Vuelidate)
```

Alternatively it is possible to import a mixin directly to components in which it will be used.

```vue
<script>
import { validationMixin } from 'vuelidate'

export default {
  mixins: [validationMixin],
  validations: { ... }
}
</script>
```

If you prefer using `require`, it can be used instead of `import` statements. This works especially great with destructuring syntax.

```js
const { validationMixin, default: Vuelidate } = require('vuelidate')
const { required, minLength } = require('vuelidate/lib/validators')
```

The browser-ready bundle is also provided in the package.

```html
<script src="vuelidate/dist/vuelidate.min.js"></script>
```
```js
// global
Vue.use(window.vuelidate.default)

// local mixin
var validationMixin = window.vuelidate.validationMixin
```
Check out the [JSFiddle example](https://jsfiddle.net/Frizi/b5v4faqf/) which uses this setup.
