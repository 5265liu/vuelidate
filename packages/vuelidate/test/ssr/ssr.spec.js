import Vue from 'vue'
import { VuelidatePlugin } from '../../src'
import { helpers } from '@vuelidate/validators'
import { createRenderer } from 'vue-server-renderer'
import VueCompositionApi from '@vue/composition-api'

Vue.use(VueCompositionApi)
Vue.use(VuelidatePlugin)

const { withParams } = helpers

const isEven = withParams({ type: 'isEven' }, (v) => {
  return v % 2 === 0
})

const mkVm = (render) =>
  new Vue({
    data: { val: 0 },
    validations: { val: { isEven } },
    render (h) {
      return render(h, this)
    }
  })

describe('SSR', () => {
  const renderer = createRenderer()

  const makeAsyncRenderTester = (vm) => new Promise((resolve, reject) => {
    renderer.renderToString(vm, (err, str) => {
      if (err) reject(str)
      else resolve(str)
    })
  })

  it('Should not throw on render when plugin is loaded', async () => {
    const vm = mkVm((h, vm) => h('div', 'hello'))
    const rendered = await makeAsyncRenderTester(vm)
    expect(rendered).toStrictEqual(expect.any(String))
  })

  it('Should not throw on render when referencing $v', async () => {
    const vm = mkVm((h, vm) => h('div', vm._s(vm.$v)))
    const rendered = await makeAsyncRenderTester(vm)
    expect(rendered).toStrictEqual(expect.any(String))
  })

  // TODO: Devise a better test
  it('Should not throw on render when referencing $v.val.isEven', async () => {
    const vm = mkVm((h, vm) => h('div', vm._s(vm.$v.val.isEven.$invalid)))
    const rendered = await makeAsyncRenderTester(vm)
    expect(rendered).toMatchSnapshot()
  })
})
