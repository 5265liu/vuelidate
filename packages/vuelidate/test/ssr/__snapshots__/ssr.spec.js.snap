// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SSR Should not throw on render when referencing $v.val.isEven 1`] = `<div>false</div>`;
